cardano-node/src/Cardano/Tracing/OrphanInstances/Network.hs

```diff
@@ -1290,14 +1514,122 @@ instance Show peerConn => ToObject (DebugPeerSelection Socket.SockAddr peerConn)
+             , "coldPeers" .= coldPeers ev
+             , "warmPeers" .= warmPeers ev
+             , "hotPeers" .= hotPeers ev
+             ]
+
+instance (Show (ClientHasAgency st), Show (ServerHasAgency st))
+  => ToJSON (PeerHasAgency pr st) where
+  toJSON (ClientAgency cha) =
+    Aeson.object [ "kind" .= String "ClientAgency"
+                 , "agency" .= show cha
+                 ]
+  toJSON (ServerAgency sha) =
+    Aeson.object [ "kind" .= String "ServerAgency"
+                 , "agency" .= show sha
+                 ]
+
+instance ToJSON ProtocolLimitFailure where
+  toJSON (ExceededSizeLimit tok) =
+    Aeson.object [ "kind" .= String "ProtocolLimitFailure"
+                 , "agency" .= toJSON tok
+                 ]
+  toJSON (ExceededTimeLimit tok) =
+    Aeson.object [ "kind" .= String "ProtocolLimitFailure"
+                 , "agency" .= toJSON tok
+                 ]
+
+instance Show vNumber => ToJSON (RefuseReason vNumber) where
+  toJSON (VersionMismatch vNumber tags) =
+    Aeson.object [ "kind" .= String "VersionMismatch"
+                 , "versionNumber" .= show vNumber
+                 , "tags" .= show tags
+                 ]
+  toJSON (HandshakeDecodeError vNumber t) =
+    Aeson.object [ "kind" .= String "HandshakeDecodeError"
+                 , "versionNumber" .= show vNumber
+                 , "text" .= show t
+                 ]
+  toJSON (Refused vNumber t) =
+    Aeson.object [ "kind" .= String "Refused"
+                 , "versionNumber" .= show vNumber
+                 , "text" .= show t
+                 ]
+
+instance Show vNumber => ToJSON (HandshakeProtocolError vNumber) where
+  toJSON (HandshakeError rvNumber) =
+    Aeson.object [ "kind" .= String "HandshakeError"
+                 , "versionNumber" .= toJSON rvNumber
+                 ]
+  toJSON (NotRecognisedVersion vNumber) =
+    Aeson.object [ "kind" .= String "NotRecognisedVersion"
+                 , "versionNumber" .= show vNumber
+                 ]
+  toJSON (InvalidServerSelection vNumber t) =
+    Aeson.object [ "kind" .= String "InvalidServerSelection"
+                 , "versionNumber" .= show vNumber
+                 , "reason" .= t
+                 ]
+
+instance Show vNumber => ToJSON (HandshakeException vNumber) where
+  toJSON (HandshakeProtocolLimit plf) =
+    Aeson.object [ "kind" .= String "HandshakeProtocolLimit"
+                 , "handshakeProtocolLimit" .= show plf
+                 ]
+  toJSON (HandshakeProtocolError err) =
+    Aeson.object [ "kind" .= String "HandshakeProtocolError"
+                 , "error" .= show err
+                 ]
+
+instance (Show versionNumber, Show agreedOptions)
+  => ToObject (ConnectionHandlerTrace versionNumber agreedOptions) where
+  toObject _verb (TrHandshakeSuccess versionNumber agreedOptions) =
+    mkObject
+      [ "kind" .= String "HandshakeSuccess"
+      , "versionNumber" .= show versionNumber
+      , "agreedOptions" .= show agreedOptions
```

Comment:
we should use a json encoding for version data

Reply:
is it safe to put a `ToJSON` constraint for that?

Reply:
yes that's ok, but we'll need to provide `ToJSON` (orhpaned) instances for `NodeToNodeVersionData` and `NodeToClientVersionData`.